src/emqx_listeners.erl

```diff
@@ -173,21 +173,35 @@ with_port({Addr, Port}, Opts = #{socket_opts := SocketOption}) ->
 restart() ->
     lists:foreach(fun restart_listener/1, emqx:get_env(listeners, [])).
 
--spec(restart_listener(listener()) -> any()).
+% emqx_listeners:find_by_id(Identifier)
+-spec(restart_listener(listener() | string() | binary()) -> ok | {error, any()}).
 restart_listener(#{proto := Proto, listen_on := ListenOn, opts := Options}) ->
-    restart_listener(Proto, ListenOn, Options).
+    restart_listener(Proto, ListenOn, Options);
+restart_listener(Identifier) ->
+    case emqx_listeners:find_by_id(Identifier) of
+        false    -> {error, {no_such_listener, Identifier}};
+        Listener -> restart_listener(Listener)
+    end.
 
--spec(restart_listener(esockd:proto(), esockd:listen_on(), [esockd:option()]) -> any()).
+-spec(restart_listener(esockd:proto(), esockd:listen_on(), [esockd:option()]) -> ok | {error, any()}).
 restart_listener(tcp, ListenOn, _Options) ->
     esockd:reopen('mqtt:tcp', ListenOn);
 restart_listener(Proto, ListenOn, _Options) when Proto == ssl; Proto == tls ->
     esockd:reopen('mqtt:ssl', ListenOn);
 restart_listener(Proto, ListenOn, Options) when Proto == http; Proto == ws ->
     _ = cowboy:stop_listener(ws_name('mqtt:ws', ListenOn)),
-    start_listener(Proto, ListenOn, Options);
+    case start_listener(Proto, ListenOn, Options) of
```

Comment:
I mean implement the 
ok/1 as a separate function and call it here like:
`ok(start_listener(Proto, ListenOn, Options));`